File: quiz-flashcard-app/backend/src/services/agents/controllerAgent.js
const { v4: uuidv4 } = require('uuid');
const { db } = require('../../config/database');
const analysisAgent = require('./analysisAgent');
const generationAgent = require('./generationAgent');
const sampleQuestionService = require('../sampleQuestionService');
const userPreferencesService = require('../userPreferencesService');

/**
 * Controller Agent - Main coordinator for multi-agent system
 *
 * Responsibilities:
 * - Coordinate between analysis and generation agents
 * - Manage workflow for question generation
 * - Handle requests and route to appropriate agents
 * - Aggregate results and maintain consistency
 */
class ControllerAgent {
  constructor() {
    this.agentName = 'controller_agent';
  }

  /**
   * Trigger analysis of sample questions
   */
  async triggerAnalysis(categoryId) {
    try {
      // Get sample questions for this category
      const sampleQuestions = sampleQuestionService.getSampleQuestionsByCategory(categoryId);

      if (!sampleQuestions || sampleQuestions.length === 0) {
        return {
          success: false,
          error: 'No sample questions found for analysis'
        };
      }

      // Log the request
      await this.logMessage(categoryId, 'user', this.agentName, 'analysis_request', {
        sampleCount: sampleQuestions.length
      });

      // Delegate to analysis agent
      const result = await analysisAgent.analyzeSampleQuestions(categoryId, sampleQuestions);

      // Log completion
      await this.logMessage(categoryId, this.agentName, 'user', 'analysis_response', {
        success: result.success,
        analyzedCount: result.analyzedCount
      });

      return result;

    } catch (error) {
      console.error('Controller Agent - Analysis Error:', error);
      return {
        success: false,
        error: error.message
      };
    }
  }

  /**
   * Generate questions with full coordination
   */
  async generateQuestions(content, options = {}) {
    const {
      categoryId,
      count = 10,
      difficulty = 'medium',
      questionType = 'multiple_choice',
      customDirections = '',
      useAnalysis = true,
      useSampleQuestions = true
    } = options;

    try {
      // Log generation request
      await this.logMessage(categoryId, 'user', this.agentName, 'generation_request', {
        count, difficulty, questionType
      });

      // Get AI insights from user preferences
      let aiInsights = null;
      if (categoryId) {
        aiInsights = userPreferencesService.getAIInsights(categoryId);
      }

      // Check if analysis exists and is recent
      let analysis = null;
      if (useAnalysis && categoryId) {
        analysis = analysisAgent.getAnalysis(categoryId);

        // If no analysis exists but we have sample questions, trigger analysis first
        if (!analysis && useSampleQuestions) {
          const sampleCount = sampleQuestionService.getSampleQuestionCount(categoryId);
          if (sampleCount > 0) {
            console.log('Controller: Triggering auto-analysis before generation...');
            await this.triggerAnalysis(categoryId);
          }
        }
      }

      // Delegate to generation agent
      const questions = await generationAgent.generateQuestions(content, {
        categoryId,
        count,
        difficulty,
        questionType,
        customDirections,
        aiInsights
      });

      // Log completion
      await this.logMessage(categoryId, this.agentName, 'user', 'generation_response', {
        success: true,
        generatedCount: questions.length
      });

      return questions;

    } catch (error) {
      console.error('Controller Agent - Generation Error:', error);
      throw error;
    }
  }

  /**
   * Get analysis status and results
   */
  getAnalysisStatus(categoryId) {
    const analysis = analysisAgent.getAnalysis(categoryId);
    const sampleCount = sampleQuestionService.getSampleQuestionCount(categoryId);
    const sampleQuestions = sampleQuestionService.getSampleQuestionsByCategory(categoryId);

    // Count by type
    const byType = {
      multiple_choice: 0,
      true_false: 0,
      written_answer: 0,
      fill_in_blank: 0
    };

    sampleQuestions.forEach(q => {
      const type = q.question_type || 'multiple_choice';
      if (byType[type] !== undefined) {
        byType[type]++;
      }
    });

    return {
      hasAnalysis: !!analysis,
      analysis: analysis,
      sampleCount,
      samplesByType: byType,
      lastUpdated: analysis?.updatedAt || null
    };
  }

  /**
   * Get agent activity log
   */
  getAgentActivity(categoryId, limit = 20) {
    const stmt = db.prepare(`
      SELECT * FROM agent_messages
      WHERE category_id = ?
      ORDER BY created_at DESC
      LIMIT ?
    `);

    const messages = stmt.all(categoryId, limit);

    return messages.map(m => ({
      id: m.id,
      from: m.from_agent,
      to: m.to_agent,
      type: m.message_type,
      payload: JSON.parse(m.payload || '{}'),
      status: m.status,
      createdAt: m.created_at,
      processedAt: m.processed_at
    }));
  }

  /**
   * Clear analysis for a category (force re-analysis)
   */
  clearAnalysis(categoryId) {
    const stmt = db.prepare('DELETE FROM ai_analysis_results WHERE category_id = ?');
    stmt.run(categoryId);

    // Log the action
    this.logMessage(categoryId, 'user', this.agentName, 'analysis_cleared', {});

    return { success: true };
  }

  /**
   * Log message to agent communication table
   */
  async logMessage(categoryId, fromAgent, toAgent, messageType, payload) {
    if (!categoryId) return;

    const id = uuidv4();
    const stmt = db.prepare(`
      INSERT INTO agent_messages (id, category_id, from_agent, to_agent, message_type, payload, status)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `);
    stmt.run(
      id,
      categoryId,
      fromAgent,
      toAgent,
      messageType,
      JSON.stringify(payload),
      'processed'
    );
  }

  /**
   * Get summary statistics for the multi-agent system
   */
  getSystemStats(categoryId) {
    const analysisStmt = db.prepare(`
      SELECT COUNT(*) as count FROM ai_analysis_results WHERE category_id = ?
    `);
    const analysisCount = analysisStmt.get(categoryId)?.count || 0;

    const messageStmt = db.prepare(`
      SELECT message_type, COUNT(*) as count
      FROM agent_messages
      WHERE category_id = ?
      GROUP BY message_type
    `);
    const messageStats = messageStmt.all(categoryId);

    const recentStmt = db.prepare(`
      SELECT * FROM agent_messages
      WHERE category_id = ?
      ORDER BY created_at DESC
      LIMIT 5
    `);
    const recentActivity = recentStmt.all(categoryId);

    return {
      analysisCount,
      messageStats: messageStats.reduce((acc, m) => {
        acc[m.message_type] = m.count;
        return acc;
      }, {}),
      recentActivity: recentActivity.map(m => ({
        type: m.message_type,
        from: m.from_agent,
        to: m.to_agent,
        createdAt: m.created_at
      }))
    };
  }
}

module.exports = new ControllerAgent();

File: quiz-flashcard-app/backend-python/agents/controller_agent.py
"""
Controller Agent - Orchestrates the multi-agent system.

Responsibilities:
- Route requests to appropriate agents
- Coordinate analysis and generation workflows
- Manage agent communication
- Track agent activity
"""
from typing import Any, Dict, List, Optional

import structlog
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from models import (
    AIAnalysisResult,
    AgentMessage as AgentMessageModel,
    Document,
    SampleQuestion,
)

from .analysis_agent import analyze_samples, clear_analysis, get_analysis_status
from .base_agent import AgentRole, BaseAgent
from .generation_agent import generate_flashcards, generate_questions
from .grading_agent import grade_answer, get_session_grades
from .handwriting_agent import (
    get_handwritten_answer,
    get_session_handwritten_answers,
    process_handwritten_answer,
    update_with_correction,
)

logger = structlog.get_logger()


class ControllerAgent(BaseAgent):
    """
    Main controller that orchestrates the multi-agent system.

    Routes requests to appropriate agents and coordinates
    complex workflows like analysis-then-generation.
    """

    def __init__(self):
        """Initialize the Controller Agent."""
        super().__init__(
            role=AgentRole.CONTROLLER,
            system_prompt="You are the controller agent coordinating AI operations.",
        )

    async def process(self, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Process a request by routing to appropriate agent.

        Args:
            input_data: Dict containing:
                - action: The action to perform
                - ... action-specific parameters

        Returns:
            Result from the appropriate agent
        """
        action = input_data.get("action")

        if not action:
            return {
                "success": False,
                "error": "No action specified",
            }

        logger.info("controller_routing", action=action)

        # Route to appropriate handler
        handlers = {
            "analyze": self._handle_analyze,
            "generate_questions": self._handle_generate_questions,
            "generate_flashcards": self._handle_generate_flashcards,
            "grade_answer": self._handle_grade_answer,
            "process_handwriting": self._handle_handwriting,
        }

        handler = handlers.get(action)
        if not handler:
            return {
                "success": False,
                "error": f"Unknown action: {action}",
            }

        return await handler(input_data)

    async def _handle_analyze(self, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """Handle analysis request."""
        db = input_data.get("db")
        category_id = input_data.get("category_id")
        force = input_data.get("force", False)

        if not db or not category_id:
            return {"success": False, "error": "Missing db or category_id"}

        return await analyze_samples(db, category_id, force)

    async def _handle_generate_questions(
        self, input_data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Handle question generation request."""
        db = input_data.get("db")
        category_id = input_data.get("category_id")
        content = input_data.get("content")
        count = input_data.get("count", 5)
        difficulty = input_data.get("difficulty", "medium")
        question_type = input_data.get("question_type", "multiple_choice")
        custom_directions = input_data.get("custom_directions", "")
        document_id = input_data.get("document_id")

        if not db or not category_id or not content:
            return {"success": False, "error": "Missing required parameters"}

        return await generate_questions(
            db=db,
            category_id=category_id,
            content=content,
            count=count,
            difficulty=difficulty,
            question_type=question_type,
            custom_directions=custom_directions,
            document_id=document_id,
        )

    async def _handle_generate_flashcards(
        self, input_data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Handle flashcard generation request."""
        db = input_data.get("db")
        category_id = input_data.get("category_id")
        content = input_data.get("content")
        count = input_data.get("count", 10)
        custom_directions = input_data.get("custom_directions", "")
        document_id = input_data.get("document_id")

        if not db or not category_id or not content:
            return {"success": False, "error": "Missing required parameters"}

        return await generate_flashcards(
            db=db,
            category_id=category_id,
            content=content,
            count=count,
            custom_directions=custom_directions,
            document_id=document_id,
        )

    async def _handle_grade_answer(self, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """Handle grading request."""
        db = input_data.get("db")
        session_id = input_data.get("session_id")
        question_id = input_data.get("question_id")
        user_answer = input_data.get("user_answer")
        use_partial_credit = input_data.get("use_partial_credit", True)

        if not all([db, session_id, question_id, user_answer]):
            return {"success": False, "error": "Missing required parameters"}

        return await grade_answer(
            db=db,
            session_id=session_id,
            question_id=question_id,
            user_answer=user_answer,
            use_partial_credit=use_partial_credit,
        )

    async def _handle_handwriting(self, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """Handle handwriting processing request."""
        db = input_data.get("db")
        session_id = input_data.get("session_id")
        question_id = input_data.get("question_id")
        image_data = input_data.get("image_data")
        file_path = input_data.get("file_path")
        original_name = input_data.get("original_name")

        if not all([db, session_id, question_id, image_data]):
            return {"success": False, "error": "Missing required parameters"}

        return await process_handwritten_answer(
            db=db,
            session_id=session_id,
            question_id=question_id,
            image_data=image_data,
            file_path=file_path or "",
            original_name=original_name or "handwritten.pdf",
        )


async def trigger_analysis(
    db: AsyncSession,
    category_id: int,
    force: bool = False,
) -> Dict[str, Any]:
    """
    Trigger analysis of sample questions.

    This is the main entry point for the analysis workflow.
    """
    # Log the action
    agent_msg = AgentMessageModel(
        category_id=category_id,
        from_agent="controller",
        to_agent="analysis_agent",
        message_type="analyze_request",
        payload={"force": force},
        status="pending",
    )
    db.add(agent_msg)

    # Run analysis
    result = await analyze_samples(db, category_id, force)

    # Update message status
    agent_msg.status = "processed"

    return result


async def get_agent_activity(
    db: AsyncSession,
    category_id: int,
    limit: int = 20,
) -> List[Dict[str, Any]]:
    """
    Get recent agent activity for a category.

    Returns the inter-agent message log.
    """
    result = await db.execute(
        select(AgentMessageModel)
        .where(AgentMessageModel.category_id == category_id)
        .order_by(AgentMessageModel.created_at.desc())
        .limit(limit)
    )
    messages = result.scalars().all()

    return [
        {
            "id": m.id,
            "from_agent": m.from_agent,
            "to_agent": m.to_agent,
            "message_type": m.message_type,
            "payload": m.payload,
            "status": m.status,
            "created_at": m.created_at.isoformat(),
            "processed_at": m.processed_at.isoformat() if m.processed_at else None,
        }
        for m in messages
    ]


async def get_system_stats(
    db: AsyncSession,
    category_id: int,
) -> Dict[str, Any]:
    """
    Get overall AI system statistics for a category.
    """
    # Count sample questions
    samples_result = await db.execute(
        select(func.count(SampleQuestion.id)).where(
            SampleQuestion.category_id == category_id
        )
    )
    sample_count = samples_result.scalar() or 0

    # Count documents
    docs_result = await db.execute(
        select(func.count(Document.id)).where(Document.category_id == category_id)
    )
    doc_count = docs_result.scalar() or 0

    # Check analysis status
    analysis_result = await db.execute(
        select(AIAnalysisResult).where(AIAnalysisResult.category_id == category_id)
    )
    analysis = analysis_result.scalar_one_or_none()

    # Count agent messages
    messages_result = await db.execute(
        select(func.count(AgentMessageModel.id)).where(
            AgentMessageModel.category_id == category_id
        )
    )
    message_count = messages_result.scalar() or 0

    return {
        "category_id": category_id,
        "sample_questions": sample_count,
        "documents": doc_count,
        "has_analysis": analysis is not None,
        "analysis_updated_at": analysis.updated_at.isoformat() if analysis else None,
        "agent_messages": message_count,
    }


async def generate_from_documents(
    db: AsyncSession,
    category_id: int,
    document_ids: Optional[List[int]] = None,
    count: int = 10,
    difficulty: str = "medium",
    question_type: str = "multiple_choice",
    custom_directions: str = "",
    chapter: str = "",
    use_rag: bool = False,
    validate: bool = False,
) -> Dict[str, Any]:
    """
    Generate questions from category documents.

    Args:
        db: Database session
        category_id: Category ID
        document_ids: Optional specific documents (None = all)
        count: Number of questions to generate
        difficulty: Difficulty level
        question_type: Question type
        custom_directions: Additional instructions
        chapter: Optional chapter/topic to tag questions with
        use_rag: Whether to use RAG for context retrieval (Phase 3)
        validate: Whether to validate and score questions (Phase 3)

    Returns:
        Generation result
    """
    # Get documents - filter by chapter if specified
    query = select(Document).where(Document.category_id == category_id)
    if document_ids:
        query = query.where(Document.id.in_(document_ids))

    # Filter by chapter if specified - only use documents from that chapter
    if chapter:
        query = query.where(Document.chapter == chapter)

    result = await db.execute(query)
    documents = result.scalars().all()

    if not documents:
        error_msg = "No documents found"
        if chapter:
            error_msg = f"No documents found for chapter '{chapter}'. Upload documents with this chapter tag or select a different chapter."
        return {
            "success": False,
            "error": error_msg,
        }

    # Combine document content (use content_text and original_name from Document model)
    combined_content = "\n\n".join(
        [f"--- {doc.original_name} ---\n{doc.content_text}" for doc in documents if doc.content_text]
    )

    if not combined_content.strip():
        return {
            "success": False,
            "error": "Documents have not been processed yet. Please wait for processing to complete.",
        }

    # Generate questions
    return await generate_questions(
        db=db,
        category_id=category_id,
        content=combined_content,
        count=count,
        difficulty=difficulty,
        question_type=question_type,
        custom_directions=custom_directions,
        document_id=documents[0].id if len(documents) == 1 else None,
        chapter=chapter,
        use_rag=use_rag,
        validate=validate,
        document_ids=document_ids or [d.id for d in documents],
    )


# Singleton instance
controller_agent = ControllerAgent()
